virtualizarr/registry.py

```diff
@@ -0,0 +1,239 @@
+
+    Raises
+    ------
+    ValueError
+        If provided Url does not contain a scheme based on [urllib.parse.urlparse][]
+    """
+    parsed = urlparse(url)
+    if not parsed.scheme:
+        raise ValueError(
+            f"Urls are expected to contain a scheme (e.g., `file://` or `s3://`), received {url} which parsed to {parsed}"
+        )
+    return UrlKey(parsed.scheme, parsed.netloc)
+
+
+class PathEntry:
+    """
+    Construct a tree of path segments starting from the root
+
+    For example the following paths:
+    * `/` => store1
+    * `/foo/bar` => store2
+
+    Would be represented by:
+    store: Some(store1)
+    children:
+      foo:
+        store: None
+        children:
+          bar:
+            store: Some(store2)
+    """
+
+    def __init__(self) -> None:
+        self.store: Optional[ObjectStore] = None
+        self.children: Dict[str, "PathEntry"] = {}
+
+    def lookup(self, to_resolve: str) -> Optional[Tuple[ObjectStore, int]]:
+        """
+        Lookup a store based on URL path
+
+        Returns the store and its path segment depth
+        """
+        current = self
+        ret = (self.store, 0) if self.store is not None else None
+        depth = 0
+
+        # Traverse the PathEntry tree to find the longest match
+        for segment in path_segments(to_resolve):
+            if segment in current.children:
+                current = current.children[segment]
+                depth += 1
+                if current.store is not None:
+                    ret = (current.store, depth)
+            else:
+                break
+
+        return ret
+
+
+class ObjectStoreRegistry:
+    def __init__(self) -> None:
+        # Mapping from UrlKey (containing scheme and netlocs) to PathEntry
+        self.map: Dict[UrlKey, PathEntry] = {}
+
+    def register(self, url: Url, store: ObjectStore) -> Optional[ObjectStore]:
+        """
+        Register a new store for the provided store URL
+
+        If a store with the same URL existed before, it is replaced and returned
+
+        Parameters
+        ----------
+        url
+            Url to registry the [ObjectStore][obstore.store.ObjectStore] under
+        store
+            [ObjectStore][obstore.store.ObjectStore] instance to register using the
+            provided url.
+
+        Returns
+        -------
+        Optional[ObjectStore]
+            The [ObjectStore][obstore.store.ObjectStore] that has been replaced, if applicable.
+
+        Examples
+        --------
+
+        ```python exec="on" source="above" session="registry-examples"
+        from obstore.store import S3Store
+        from virtualizarr.registry import ObjectStoreRegistry
+
+        reg = ObjectStoreRegistry()
+        orig_store = S3Store(bucket="my-bucket-1", prefix="orig-path")
+        reg.register("s3://my-bucket-1", orig_store)
+
+        new_store = S3Store(bucket="my-bucket-1", prefix="updated-path")
+        store = reg.register("s3://my-bucket-1", new_store)
```

Comment:
We discussed yesterday that this API goes against a common python convention, that functions should either modify in-place or have a return value, but not both.

References:
- https://github.com/pydata/xarray/issues/9196
- https://stackoverflow.com/questions/58269413/why-in-principle-an-in-place-modifying-method-should-return-none